### PR TITLE
spectre-cli: init at unstable-2022-02-05

### PR DESCRIPTION
**Pull request by emmabastas**
###### Description of changes

A CLI wrapper for the "spectre algorithmn"
homepage: https://spectre.app/

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by NOTtheMessiah** [comment]
> ###### Description of changes
> 
> A CLI wrapper for the "spectre algorithmn" homepage: https://spectre.app/
> ###### Things done
> 
>     * Built on platform(s)
>       
>       * [x]  x86_64-linux
>       * [ ]  aarch64-linux
>       * [ ]  x86_64-darwin
>       * [ ]  aarch64-darwin
> 
>     * [ ]  For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
> 
>     * [x]  Tested, as applicable:
>       
>       * [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
> 
>     * [x]  Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
> 
>     * [x]  Tested basic functionality of all binary files (usually in `./result/bin/`)
> 
>     * [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
>       
>       * [ ]  (Package updates) Added a release notes entry if the change is major or breaking
>       * [ ]  (Module updates) Added a release notes entry if the change is significant
>       * [ ]  (Module addition) Added a release notes entry if adding a new NixOS module
>       * [ ]  (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
> 
>     * [ ]  Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

File: pkgs/tools/security/spectre-cli/default.nix [hunk 1]
```diff
@@ -0,0 +1,67 @@
+{ lib
+, gcc11Stdenv
+, fetchFromGitLab
+, cmake
+, libsodium
+, json_c
+, ncurses
+, libxml2
+, jq
+, ...
```
Comment (FliegendeWurst) [inline pkgs/tools/security/spectre-cli/default.nix:10]:
```suggestion
```

File: pkgs/tools/security/spectre-cli/default.nix [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  version = "unstable-2021-02-05";
+
+  src = fetchFromGitLab {
+    owner = "spectre.app";
+    repo = "cli";
+    rev = "a5e7aab28f44b90e5bd1204126339a81f64942d2";
+    sha256 = "1kda4jj7sh940fpy9cqfgzmwjlwnkr75lqgm3addpzpblnw6vvwr";
+    fetchSubmodules = true;
+    deepClone = true;
+    leaveDotGit = true;
```
Comment (FliegendeWurst) [inline pkgs/tools/security/spectre-cli/default.nix:24]:
```suggestion
    sha256 = "1hp4l1rhg7bzgx0hcai08rvcy6l9645sfngy2cr96l1bpypcld5i";
    fetchSubmodules = true;
```
Is there a particular reason the build needs the `.git` directory?

File: pkgs/tools/security/spectre-cli/default.nix [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+  cmakeFlags = [
+    "-DBUILD_SPECTRE_TESTS=ON"
+  ];
+
+  preConfigure = ''
+    # The default buildPhase wants to create a ´build´ dir so we rename the build script to stop conflicts.
+    mv build build.sh
+  '';
+
```
Comment (FliegendeWurst) [inline pkgs/tools/security/spectre-cli/default.nix:47]:
```suggestion
  doCheck = true;
```

File: pkgs/tools/security/spectre-cli/default.nix [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  checkPhase = ''
+    mv ../spectre-cli-tests ../spectre_tests.xml ./
+    ./spectre-tests
```
Comment (FliegendeWurst) [inline pkgs/tools/security/spectre-cli/default.nix:50]:
```suggestion
    patchShebangs spectre-cli-tests
    export HOME=$(mktemp -d)

    ./spectre-tests
```

File: pkgs/tools/security/spectre-cli/default.nix [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  installPhase = ''
+    mkdir -p $out/bin
+    mv spectre $out/bin
+  '';
+
+  meta = with lib; {
+    description = "A stateless cryptographic identity algorithm";
+    homepage = "https://spectre.app";
+    license = licenses.gpl3;
+    maintainers = [ lib.maintainers.emmabastas ];
```
Comment (FliegendeWurst) [inline pkgs/tools/security/spectre-cli/default.nix:63]:
```suggestion
    license = licenses.gpl3Plus;
    maintainers = with maintainers; [ emmabastas ];
```

**Review by FliegendeWurst — COMMENTED** [review]
Welcome to nixpkgs :)

Your PR looks pretty good already. I left a few suggestions in my review. Please also change your second commit message to conform to https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes

File: pkgs/tools/security/spectre-cli/default.nix [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+, cmake
+, libsodium
+, json_c
+, ncurses
+, libxml2
+, jq
+, ...
+}:
+
+gcc11Stdenv.mkDerivation {
```
Comment (FliegendeWurst) [inline pkgs/tools/security/spectre-cli/default.nix:13]:
```suggestion
stdenv.mkDerivation {
```
GCC 11 is the default for {x86_64,aarch64}-{darwin,linux}.

File: pkgs/tools/security/spectre-cli/default.nix (hunk 2 above)
Comment (emmabastas) [reply pkgs/tools/security/spectre-cli/default.nix:24]: No reason, forgot to remove them. `fetchSubmodules` is still needed as you noticed :-)

**Comment by emmabastas** [comment]
Thanks for the welcome :-)
And for the suggestions, I've incorporated them all now. I noticed that the test-suite [fails on ARM64](https://github.com/NixOS/nixpkgs/pull/174811/checks?check_run_id=6623982987) and have opened [this issue](https://gitlab.com/spectre.app/cli/-/issues/27) upstream.

Unsure what a good course of action is now.

**Comment by FliegendeWurst** [comment]
> I noticed that the test-suite [fails on ARM64](https://github.com/NixOS/nixpkgs/pull/174811/checks?check_run_id=6623982987) and have opened [this issue](https://gitlab.com/spectre.app/cli/-/issues/27) upstream.

You should set `platforms = [ "x86_64-linux" "x86_64-darwin" "aarch64-darwin" ];`. Once the issue is fixed upstream you can update the package and specify `platforms = platforms.all;`.

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -35405,4 +35405,6 @@ with pkgs;
   honeyvent = callPackage ../servers/tracing/honeycomb/honeyvent { };
 
   mictray = callPackage ../tools/audio/mictray { };
+
+  spectre-cli = callPackage ../tools/security/spectre-cli { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:35409]: Please sort this into a fitting category and sort there alphabetically.

File: pkgs/tools/security/spectre-cli/default.nix [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    json_c
+    ncurses
+  ];
+
+  cmakeFlags = [
+    "-DBUILD_SPECTRE_TESTS=ON"
+  ];
+
+  preConfigure = ''
+   echo "unstable-2021-02-05" > VERSION
```
Comment (SuperSandro2000) [inline pkgs/tools/security/spectre-cli/default.nix:41]:
```suggestion
   echo "${version}" > VERSION
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/tools/security/spectre-cli/default.nix [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+    # The default buildPhase wants to create a ´build´ dir so we rename the build script to stop conflicts.
+    mv build build.sh
+  '';
+  doCheck = true;
```
Comment (FliegendeWurst) [inline pkgs/tools/security/spectre-cli/default.nix:46]:
```suggestion
  doCheck = !(stdenv.isLinux && stdenv.isAarch64);
```
Some tests are "expected" to fail on ARM64: https://gitlab.com/spectre.app/cli/-/issues/27#note_962950844

**Comment by emmabastas** [comment]
Thank u @FliegendeWurst and @SuperSandro2000  for helping me make the changes need for a merge. Hopefully I'll be able to pay it forward some day :-)

**Pull request merged**